README.rst
====
twod
====

``twod`` is a daemon that updates your TwoDNS_ host IP entries.


configuration
=============

``twod`` will look for a configuration file in ``/etc/twod/twodrc``. Optionally
you can tell ``twod`` to use a specified configuration file instead by using
the ``-c`` parameter.

Example config::

    [general]
    user      = username@example.com
    token     = token
    host_url  = https://api.twodns.de/hosts/myexamplehost
    interval  = 3600
    timeout   = 16
    redirects = 2

    [ip_service]
    mode      = random
    ip_urls   = https://icanhazip.com https://ipinfo.io/ip

    [logging]
    level     = WARNING


installation
============

Gentoo GNU/Linux
^^^^^^^^^^^^^^^^

1. Install ``twod`` from `my little overlay <https://github.com/twisted-pear/my-little-overlay>`_.

2. Copy ``/usr/share/doc/twod-<version>/examples/twodrc.example`` to
   ``/etc/twodrc`` and change the settings according to your setup.


usage
=====

After installing ``twod`` you can control it like any other service:

    $ /etc/init.d/twod start

To run the daemon manually you can just call the ``twod`` binary:

    $ twod


documentation
=============

You can find detailed documentation at
`twod's Read the Docs page <https://twod.readthedocs.org/en/latest/>`_,
powered by Sphinx_.


tests
=====

Tests can be run by simply installing and invoking tox:

   $ tox



.. _TwoDNS: https://www.twodns.de
.. _my_little_overlay: https://github.com/twisted-pear/my-little-overlay
.. _Sphinx: http://sphinx-doc.org


docs/changelog.rst
.. _changelog:

Changelog
=========

0.5.1
-----

* Fix packaging issues. [`#8 <https://github.com/tablet-mode/twod/issues/8>`_]

0.5.0
-----

* Add python 3 compatibility. [`#7 <https://github.com/tablet-mode/twod/issues/7>`_]

* Tidy up a bit.

0.4.0
-----

* Add config setting for maximum number of redirects to follow on HTTP
  requests.

* Update man pages.

* Add CLI argument for pidfile location.

* Make interval, timeout, redirects, ip_mode and loglevel settings optional.

* Never exit on errors caused by HTTP requests.

0.3.2
-----

* Add config setting for HTTP timeout.

* Fix terminology for token setting. [`#1 <https://github.com/tablet-mode/twod/issues/1>`_]

* Catch connection and response timeouts. [`#2 <https://github.com/tablet-mode/twod/issues/2>`_]

* Normalise version string in setup.py.

* Add IP validation.

* Get rid of "Unexpected error" exceptions.

0.3.1
-----

* Add python dependencies to setup.py.

* Implement simpler version handling.

* Fix capitalisation of `TwoDNS <https://twodns.de>`_.

* Improve error handling for config parsing.


tests/conftest.py
"""Fixtures for twod."""

import pytest


# Config fixtures
@pytest.fixture
def empty_config(tmpdir):
    """Invalid configuration."""
    f = tmpdir.join("twodrc")
    f.write("setting =")
    return tmpdir


@pytest.fixture
def valid_config(tmpdir):
    """Valid configuration."""
    f = tmpdir.join("twodrc")
    f.write("""
[general]
user     = username@example.com
token = token
host_url = https://api.twodns.de/hosts/example.dd-dns.de
interval = 9000
timeout = 9000

[ip_service]
mode     = random
ip_urls  = https://icanhazip.com https://ipinfo.io/ip

[logging]
level    = WARNING
""")
    return tmpdir


@pytest.fixture
def valid_config_mode_rr(tmpdir):
    """Valid configuration."""
    f = tmpdir.join("twodrc")
    f.write("""
[general]
user     = username@example.com
token = token
host_url = https://api.twodns.de/hosts/example.dd-dns.de
interval = 9000
timeout = 9000

[ip_service]
mode     = round_robin
ip_urls  = https://nr_one https://nr_two https://nr_three

[logging]
level    = WARNING
""")
    return tmpdir


@pytest.fixture
def invalid_host_config(tmpdir):
    """Invalid configuration."""
    f = tmpdir.join("twodrc")
    f.write("""
[general]
user     = username@example.com
token = token
host_url = https://127.0.0.1:57357
interval = 9001
timeout = 9000

[ip_service]
mode     = random
ip_urls  = https://127.0.0.2:57357

[logging]
level    = WARNING
""")
    return tmpdir


@pytest.fixture
def invalid_url_config(tmpdir):
    """Invalid configuration."""
    f = tmpdir.join("twodrc")
    f.write("""
[general]
user     = username@example.com
token = token
host_url = invalid_url
interval = 9001
timeout = 9000

[ip_service]
mode     = random
ip_urls  = invalid_url

[logging]
level    = WARNING
""")
    return tmpdir


@pytest.fixture
def invalid_mode_config(tmpdir):
    """Invalid configuration."""
    f = tmpdir.join("twodrc")
    f.write("""
[general]
user     = username@example.com
token = token
host_url = https://127.0.0.1:57357
interval = 9001
timeout = 9000

[ip_service]
mode     = invalid_mode
ip_urls  = https://127.0.0.2:57357

[logging]
level    = WARNING
""")
    return tmpdir


@pytest.fixture
def missing_username_config(tmpdir):
    """Valid configuration."""
    f = tmpdir.join("twodrc")
    f.write("""
[general]
token = token
host_url = https://api.twodns.de/hosts/example.dd-dns.de
interval = 9000
timeout = 9000

[ip_service]
mode     = random
ip_urls  = https://icanhazip.com https://ipinfo.io/ip

[logging]
level    = WARNING
""")
    return tmpdir


@pytest.fixture
def missing_section_config(tmpdir):
    """Valid configuration."""
    f = tmpdir.join("twodrc")
    f.write("""
[general]
user     = username@example.com
token = token
host_url = https://api.twodns.de/hosts/example.dd-dns.de
interval = 9000
timeout = 9000

[logging]
level    = WARNING
""")
    return tmpdir


@pytest.fixture
def valid_config_path(valid_config):
    """Path to valid config."""
    pathstring = ('{dir}/{base}/twodrc'.format(dir=valid_config.dirname,
                                               base=valid_config.basename))
    return pathstring


@pytest.fixture
def valid_config_mode_rr_path(valid_config_mode_rr):
    """Path to valid config with mode set to ``round_robin``."""
    pathstring = ('{dir}/{base}/twodrc'.format(
        dir=valid_config_mode_rr.dirname, base=valid_config_mode_rr.basename))
    return pathstring


@pytest.fixture
def invalid_host_config_path(invalid_host_config):
    """Path to valid config with invalid host entry."""
    pathstring = ('{dir}/{base}/twodrc'.format(
        dir=invalid_host_config.dirname, base=invalid_host_config.basename))
    return pathstring


@pytest.fixture
def invalid_url_config_path(invalid_url_config):
    """Path to valid config with invalid url entry."""
    pathstring = ('{dir}/{base}/twodrc'.format(
        dir=invalid_url_config.dirname, base=invalid_url_config.basename))
    return pathstring


@pytest.fixture
def invalid_mode_config_path(invalid_mode_config):
    """Path to valid config with invalid mode entry."""
    pathstring = ('{dir}/{base}/twodrc'.format(
        dir=invalid_mode_config.dirname, base=invalid_mode_config.basename))
    return pathstring


@pytest.fixture
def missing_username_config_path(missing_username_config):
    """Path to config with missing username."""
    pathstring = ('{dir}/{base}/twodrc'.format(
        dir=missing_username_config.dirname,
        base=missing_username_config.basename))
    return pathstring


@pytest.fixture
def missing_section_config_path(missing_section_config):
    """Path to config without ip_service section."""
    pathstring = ('{dir}/{base}/twodrc'.format(
        dir=missing_section_config.dirname,
        base=missing_section_config.basename))
    return pathstring


# Data fixtures
@pytest.fixture
def twodns_response():
    """Response fixture."""
    response = (u'{"activate_wildcard":false,"ttl":60,'
                '"fqdn":"example.dd-dns.de","ip_address":"127.0.0.1",'
                '"url":"https://api.twodns.de/hosts/example.dd-dns.de"}')
    return response


tests/test_cli.py
"""Tests for the CLI of twod."""

import pytest
import mock

from twod.twod import main
from twod._version import __version__


class TestCLI:
    """Test CLI."""

    def test_cli(self, capsys, monkeypatch):
        """Test CLI."""
        monkeypatch.setattr('sys.argv', ['twod.py', '-h'])
        with pytest.raises(SystemExit):
            main()
        out, err = capsys.readouterr()
        assert "usage:" in out

    def test_config_argument_no_sections(self, capsys, monkeypatch,
                                         empty_config):
        """Test CLI with --config argument and no sections."""
        monkeypatch.setattr('sys.argv', ['twod.py', '-c',
                            '{dir}/{base}/twodrc'.format(
                                dir=empty_config.dirname,
                                base=empty_config.basename)])
        with pytest.raises(SystemExit):
            main()
        out, err = capsys.readouterr()
        assert "File contains no section headers" in err

    def test_invalid_config_path(self, capsys, monkeypatch):
        """Test CLI with --config argument and invalid config path."""
        monkeypatch.setattr('sys.argv', ['twod.py', '-c', 'invalid_path'])
        with pytest.raises(SystemExit):
            main()
        out, err = capsys.readouterr()
        assert "'invalid_path' is not a file" in err

    # Exit with message ``DAEMON`` if daemonisation is attempted.
    @mock.patch('twod.twod.DaemonContext', side_effect=SystemExit("DAEMON"))
    @mock.patch('twod.twod._Data')
    @mock.patch('twod.twod.Session.get')
    # If we reach the ``sleep`` statement we already completed one update
    # cycle so we can exit there.
    @mock.patch('twod.twod.sleep', side_effect=SystemExit("TEST DONE"))
    def test_no_detach(self, mock_sleep, mock_get, mock_data, mock_daemon,
                       capsys, caplog, monkeypatch, valid_config_path):
        """Test --no-detach argument."""
        monkeypatch.setattr('sys.argv', ['twod.py', '-D', '-c',
                            valid_config_path])
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.1"}')
        mock_get.return_value = MyMock

        with pytest.raises(SystemExit) as e:
            main()
        assert "TEST DONE" in str(e)

    def test_version(self, capsys, monkeypatch):
        """Test CLI."""
        monkeypatch.setattr('sys.argv', ['twod.py', '-V'])
        with pytest.raises(SystemExit):
            main()
        out, err = capsys.readouterr()
        assert "twod {}".format(__version__) in out + err


tests/test_data.py
"""Tests for twod's main function."""

import mock
from requests import exceptions

from twod.twod import Twod, _Data


class TestData:
    """Test main function."""

    @mock.patch('twod.twod.Session.get')
    def test_get_rec_ip(self, mock_get, capsys, valid_config_path):
        """Test retrieval of recorded IP."""
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.2"}')
        mock_get.return_value = MyMock
        cls = Twod(valid_config_path)
        data = _Data(cls.conf)

        assert cls.interval == 9000
        assert data.rec_ip == '127.0.0.2'

    def test_get_rec_ip_invalid_host(self, capsys, caplog,
                                     invalid_host_config_path):
        """Test config parsing with invalid host."""
        cls = Twod(invalid_host_config_path)
        data = _Data(cls.conf)

        assert data._get_rec_ip() is False
        assert cls.interval == 9001
        assert "error while fetching ip from twodns" in (
            caplog.text.lower())

    @mock.patch('twod.twod.Session.get')
    def test_get_ext_ip(self, mock_get, capsys, caplog,
                        valid_config_path):
        """Test retrieval of external IP."""
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.2"}')
        MyMock2 = mock.Mock(text="127.0.0.3")
        mock_get.return_value = MyMock
        cls = Twod(valid_config_path)
        data = _Data(cls.conf)

        assert cls.interval == 9000
        assert data.rec_ip == '127.0.0.2'

        mock_get.return_value = MyMock2
        assert data._get_ext_ip() == '127.0.0.3'

    def test_get_ext_ip_invalid_host(self, capsys, caplog,
                                     invalid_host_config_path):
        """Test config parsing with invalid IP service URLs."""
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.2"}')
        patcher = mock.patch('twod.twod.Session.get')
        my_mock = patcher.start()
        my_mock.return_value = MyMock
        cls = Twod(invalid_host_config_path)
        data = _Data(cls.conf)
        assert data.rec_ip == "127.0.0.2"

        patcher.stop()
        assert data._get_ext_ip() is False
        assert "error while fetching external ip" in (
            caplog.text.lower())

    @mock.patch('twod.twod.Session.get')
    def test_get_ext_ip_rr(self, mock_get, capsys, caplog,
                           valid_config_mode_rr_path):
        """Test round robin URL selection mode."""
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.2"}')
        mock_get.return_value = MyMock
        cls = Twod(valid_config_mode_rr_path)
        data = _Data(cls.conf)

        # test round robin service URL selection
        assert data._get_service_url() == 'https://nr_one'
        assert data._get_service_url() == 'https://nr_two'
        assert data._get_service_url() == 'https://nr_three'
        assert data._get_service_url() == 'https://nr_one'
        assert data._get_service_url() == 'https://nr_two'
        assert data._get_service_url() == 'https://nr_three'
        assert data._get_service_url() == 'https://nr_one'

    @mock.patch('twod.twod.Session.get')
    def test_check(self, mock_get, capsys, caplog,
                   valid_config_path):
        """Test IP comparison."""
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.2"}')
        MyMock2 = mock.Mock(text="127.0.0.3")
        MyMock3 = mock.Mock(text="127.0.0.2")
        mock_get.return_value = MyMock
        cls = Twod(valid_config_path)
        data = _Data(cls.conf)

        assert cls.interval == 9000
        assert data.rec_ip == '127.0.0.2'

        # external and recorded IP are different, return external IP
        mock_get.return_value = MyMock2
        assert data._get_ext_ip() == '127.0.0.3'
        assert data._check_ip() == '127.0.0.3'

        # external and recorded IP are the same, return False
        mock_get.return_value = MyMock3
        assert data._check_ip() is False

    @mock.patch('twod.twod.Session.get')
    @mock.patch('twod.twod.Session.put')
    def test_update(self, mock_put, mock_get, capsys, caplog,
                    valid_config_path):
        """Test IP update."""
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.2"}')
        MyMock2 = mock.Mock(status_code=200)
        mock_get.return_value = MyMock
        mock_put.return_value = MyMock2
        cls = Twod(valid_config_path)
        data = _Data(cls.conf)

        data._update_ip('127.0.0.3')
        assert data.rec_ip == '127.0.0.3'

    @mock.patch('twod.twod.Session.get')
    @mock.patch('twod.twod.Session.put')
    def test_update_fail(self, mock_put, mock_get, capsys, caplog,
                         valid_config_path):
        """Test IP update failure."""
        MyMock = mock.Mock(text=u'{"ip_address": "127.0.0.2"}')

        def http_error(*args, **kwargs):
            raise exceptions.HTTPError("Service Unavailable")

        MyMock2 = mock.Mock(status_code=503, side_effect=http_error)
        mock_get.return_value = MyMock
        mock_put.side_effect = MyMock2
        cls = Twod(valid_config_path)
        data = _Data(cls.conf)

        data._update_ip('127.0.0.3')
        assert data.rec_ip == '127.0.0.2'
        assert "Error while updating IP" in caplog.text


twod/twod.py
#!/usr/bin/env python

"""twod TwoDNS host IP updater daemon.

twod is a client for the TwoDNS dynamic DNS service.

Copyright (C) 2014 Thomas Kager <tablet-mode AT monochromatic DOT cc>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see [http://www.gnu.org/licenses/].

"""

from __future__ import absolute_import

import logging.handlers
import logging.config

from argparse import ArgumentParser
from configparser import (SafeConfigParser, MissingSectionHeaderError,
                          NoSectionError, NoOptionError)
from json import dumps, loads
from lockfile.pidlockfile import PIDLockFile
from os import access, path, W_OK, X_OK
from random import randint
from re import match
from socket import inet_pton, error as socket_error, AF_INET, AF_INET6
from time import sleep

from daemon import DaemonContext
from requests import exceptions, Session

from twod._version import __version__


class _ServiceGenerator(object):
    """Select service URL depending on mode."""

    def __init__(self, services, mode):
        self.services = services
        self.mode = mode
        self.cur = -1

    def __iter__(self):
        return self

    def next(self):
        self.cur = self.cur + 1
        if self.mode == 'round_robin':
            if self.cur < len(self.services):
                service = self.services[self.cur]
            else:
                self.cur = 0
                service = self.services[self.cur]
            return service
        elif self.mode == 'random':
            self.cur = randint(0, (len(self.services) - 1))
            service = self.services[self.cur]
            return service


class _Data(object):
    """This is where the fun begins."""

    def __init__(self, conf):
        self.log = logging.getLogger('twod')
        self.ident = (conf['user'], conf['token'])
        self.url = conf['url']
        self.timeout = conf['timeout']
        self.redirects = conf['redirects']
        self.gen = _ServiceGenerator(conf['ip_url'].split(' '),
                                     conf['ip_mode'])
        self.rec_ip = self._get_rec_ip()

    def _validate_ip(self, ip, families=[4, 6]):
        """Validate textual IP address representation.

        Returns passed IP string if valid.
        Returns False on failure.

        """
        for family in families:
            try:
                if family == 4:
                    inet_pton(AF_INET, ip)
                elif family == 6:
                    inet_pton(AF_INET6, ip)
            except (socket_error, UnicodeEncodeError):
                # This is fine
                pass
            else:
                return ip
        return False

    def _get_service_url(self):
        """Get next URL from service generator."""
        return self.gen.next()

    def _get_ext_ip(self):
        """Get external IP.

        Returns external IP as string.
        Returns False on failure.

        """
        self.log.debug("Fetching external IP...")
        try:
            with Session() as s:
                s.max_redirects = self.redirects
                ip_request = s.get(self._get_service_url(), verify=True,
                                   timeout=self.timeout)
            ip_request.raise_for_status()
        except (exceptions.ConnectionError, exceptions.HTTPError) as e:
            self.log.warning("Error while fetching external IP: %s" % e)
            return False
        except exceptions.Timeout:
            self.log.warning("Failed to fetch external IP: Server did not "
                             "respond within %s seconds" % self.timeout)
            return False
        except exceptions.TooManyRedirects:
            self.log.warning("Failed to fetch external IP: "
                             "Too many redirects")
            return False
        except Exception as e:
            self.log.error("Unexpected error while fetching external IP "
                           ", retrying at next interval: %s" % e)
            return False
        else:
            ip = ip_request.text.rstrip()
            if not self._validate_ip(ip):
                self.log.warning("External IP discovery returned invalid IP")
            else:
                return ip

    def _get_rec_ip(self):
        """Get IP stored by TwoDNS.

        Returns IP as string. Returns False on failure.

        """
        self.log.debug("Fetching TwoDNS IP...")
        try:
            with Session() as s:
                s.max_redirects = self.redirects
                rec_request = s.get(
                    self.url, auth=self.ident, verify=True,
                    timeout=self.timeout)
            rec_request.raise_for_status()
        except (exceptions.ConnectionError, exceptions.HTTPError) as e:
            self.log.warning("Error while fetching IP from TwoDNS: %s" % e)
            return False
        except exceptions.Timeout:
            self.log.warning("Failed to fetch TwoDNS IP: Server did not "
                             "respond within %s seconds" % self.timeout)
            return False
        except exceptions.TooManyRedirects:
            self.log.warning("Failed to fetch TwoDNS IP: Too many redirects")
            return False
        except Exception as e:
            self.log.error("Unexpected error while fetching TwoDNS IP, "
                           "retrying at next interval: %s" % e)
            return False
        else:
            rec_json = loads(rec_request.text)
            ip = rec_json['ip_address']
            if not self._validate_ip(ip):
                self.log.warning("TwoDNS returned invalid IP")
            else:
                return ip

    def _check_ip(self):
        """Check if external IP matches recorded IP.

        Returns external IP as string if IPs differ. Returns False if the IPs
        match or an error occured.

        """
        self.log.debug("Checking if recorded IP matches current IP...")
        ext_ip = self._get_ext_ip()
        # something went wrong while fetching external IP but it's possible to
        # continue
        if not ext_ip:
            return False

        rec_ip = self.rec_ip
        if ext_ip == rec_ip:
            self.log.debug("IP has not changed.")
            return False
        else:
            return ext_ip

    def _update_ip(self, new_ip):
        """Update IP stored at TwoDNS."""
        self.log.debug("Updating recorded IP...")
        payload = {"ip_address": new_ip}
        try:
            with Session() as s:
                s.max_redirects = self.redirects
                rq = s.put(
                    self.url, auth=self.ident, data=dumps(payload),
                    verify=True, timeout=self.timeout)
            rq.raise_for_status()
        except (exceptions.ConnectionError, exceptions.HTTPError) as e:
            self.log.warning("Error while updating IP: %s" % e)
        except exceptions.Timeout:
            self.log.warning("Failed to update IP: Server did not respond "
                             "within %s seconds" % self.timeout)
        except exceptions.TooManyRedirects:
            self.log.warning("Failed to update IP: Too many redirects")
        except Exception as e:
            self.log.error("Unexpected error while updating TwoDNS IP, "
                           "retrying at next interval: %s" % e)
        else:
            self.log.info("IP changed to %s." % new_ip)
            self.rec_ip = new_ip


class Twod(object):
    """Twod class."""

    def __init__(self, config_path='/etc/twod/twodrc'):
        """Initialisation.

        * Setup logging
        * Read configuration
        * Initialise Data class

        """
        self._setup_logger()
        conf = self._read_config(config_path)
        self._setup_logger(conf['loglevel'])
        self.interval = conf['interval']
        self.conf = conf

    def _is_url(self, url):
        if not match(r'http(s)?://', url):
            raise ValueError(
                "Invalid URL: '%s' - has to start with 'http(s)'" % url)
        return url

    def _is_mode(self, mode):
        if mode not in ('random', 'round_robin'):
            raise ValueError("Invalid mode: '%s'" % mode)
        return mode

    def _setup_logger(self, level='WARNING'):
        """Setup logging."""
        logging.config.dictConfig({
            'version': 1,
            'disable_existing_loggers': False,
            'formatters': {
                'daemon': {
                    'format': '%(asctime)s %(module)s[%(process)d]: '
                              '%(message)s'
                },
            },
            'handlers': {
                'syslog': {
                    'formatter': 'daemon',
                    'class': 'logging.handlers.SysLogHandler',
                    'address': '/dev/log',
                    'level': 'DEBUG',
                },
                'stderr': {
                    'formatter': 'daemon',
                    'class': 'logging.StreamHandler',
                    'stream': 'ext://sys.stderr',
                    'level': 'DEBUG',
                },
            },
            'loggers': {
                'twod': {
                    'handlers': ['syslog', 'stderr'],
                    'level': level,
                    'propagate': True
                }
            }
        })
        self.log = logging.getLogger('twod')

    def _read_config(self, config_path):
        """Read config.

        Exit on invalid config.

        """
        self.log.debug("Reading config...")
        conf = {}
        defaults = {
            'interval': '3600',
            'timeout': '16',
            'redirects': '2',
            'ip_mode': 'random',
            'loglevel': 'WARNING',
        }
        config = SafeConfigParser(defaults=defaults)
        try:
            # Check if config is even readable
            f = open(path.expanduser(config_path), 'r')

            # Read config
            config.readfp(f)
            f.close()

            conf['user'] = config.get('general', 'user')
            conf['token'] = config.get('general', 'token')
            conf['url'] = self._is_url(config.get('general', 'host_url'))
            conf['interval'] = config.getfloat('general', 'interval')
            conf['timeout'] = config.getfloat('general', 'timeout')
            conf['redirects'] = config.getint('general', 'redirects')
            conf['ip_mode'] = self._is_mode(config.get('ip_service', 'mode'))
            conf['ip_url'] = self._is_url(config.get('ip_service', 'ip_urls'))
            conf['loglevel'] = config.get('logging', 'level',
                                          fallback='WARNING')
        except (MissingSectionHeaderError, NoSectionError, NoOptionError,
                ValueError, IOError) as e:
            self.log.critical("Configuration error: %s" % e)
            exit(1)
        return conf

    def run(self):
        """Main loop."""
        data = _Data(self.conf)
        while(True):
            changed_ip = data._check_ip()
            if changed_ip:
                data._update_ip(changed_ip)
            sleep(self.interval)


def main():
    """Main function."""
    parser = ArgumentParser()
    parser.add_argument('-c', '--config', metavar='FILE',
                        help="load configuration from FILE")
    parser.add_argument('-p', '--pidfile', metavar='FILE',
                        help="use FILE as pidfile")
    parser.add_argument('-D', '--no-detach', dest='nodetach',
                        action='store_true',
                        help="do not detach from console")
    parser.add_argument('-V', '--version', action='version',
                        version='twod ' + __version__)
    args = parser.parse_args()

    if args.config and not path.isfile(path.expanduser(args.config)):
        parser.error("'%s' is not a file" % args.config)

    twod = Twod(args.config) if args.config else Twod()
    if args.nodetach:
        twod.run()
    else:
        pidfile = '/var/run/twod.pid'
        if args.pidfile:
            pidfile = args.pidfile
        twod.log.debug("Moving to background...")

        # Try to catch issues with pidfile before detaching, but as late as
        # possible to avoid race conditions
        if not access(path.dirname(pidfile), W_OK | X_OK):
            twod.log.critical("Unable to write pidfile")
            exit(1)
        with DaemonContext(pidfile=PIDLockFile(pidfile)):
            twod.run()


if __name__ == '__main__':
    main()
